component/storage/leveldb/leveldb.go

```diff
@@ -0,0 +1,503 @@
+)
+
+// Provider leveldb implementation of storage.Provider interface.
+type Provider struct {
+	dbPath string
+	dbs    map[string]*store
+	lock   sync.RWMutex
+}
+
+type closer func(storeName string)
+
+type tagMapping map[string]map[string]struct{} // map[TagName](Set of database Keys)
+
+type dbEntry struct {
+	Value []byte        `json:"value,omitempty"`
+	Tags  []storage.Tag `json:"tags,omitempty"`
+}
+
+// NewProvider instantiates Provider.
+func NewProvider(dbPath string) *Provider {
+	return &Provider{dbs: make(map[string]*store), dbPath: dbPath}
+}
+
+// OpenStore opens and returns a store for given name space.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	if name == "" {
+		return nil, errors.New("store name cannot be blank")
+	}
+
+	name = strings.ToLower(name)
+
+	store := p.getLeveldbStore(name)
+	if store == nil {
+		return p.newLeveldbStore(name)
+	}
+
+	return store, nil
+}
+
+// SetStoreConfig isn't needed for LevelDB. For consistency with other store implementations, it saves
+// the store config for later retrieval.
+func (p *Provider) SetStoreConfig(name string, config storage.StoreConfiguration) error {
+	name = strings.ToLower(name)
+
+	openStore, ok := p.dbs[name]
+	if !ok {
+		return storage.ErrStoreNotFound
+	}
+
+	configBytes, err := json.Marshal(config)
+	if err != nil {
+		return fmt.Errorf("failed to marshal store configuration: %w", err)
+	}
+
+	err = openStore.Put(storeConfigKey, configBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put store store configuration: %w", err)
+	}
+
+	return nil
+}
+
+// GetStoreConfig returns the current store configuration.
+func (p *Provider) GetStoreConfig(name string) (storage.StoreConfiguration, error) {
+	name = strings.ToLower(name)
+
+	openStore, ok := p.dbs[name]
+	if !ok {
+		return storage.StoreConfiguration{}, storage.ErrStoreNotFound
+	}
+
+	storeConfigBytes, err := openStore.Get(storeConfigKey)
+	if err != nil {
+		return storage.StoreConfiguration{},
+			fmt.Errorf(`failed to get store configuration for "%s": %w`, name, err)
+	}
+
+	var storeConfig storage.StoreConfiguration
+
+	err = json.Unmarshal(storeConfigBytes, &storeConfig)
+	if err != nil {
+		return storage.StoreConfiguration{}, fmt.Errorf("failed to unmarshal store configuration: %w", err)
+	}
+
+	return storeConfig, nil
+}
+
+// GetOpenStores is not implemented.
+func (p *Provider) GetOpenStores() []storage.Store {
+	panic("implement me")
+}
+
+// Close closes all stores created under this store provider.
+func (p *Provider) Close() error {
+	p.lock.RLock()
+
+	openStoresSnapshot := make([]*store, len(p.dbs))
+
+	var counter int
+
+	for _, openStore := range p.dbs {
+		openStoresSnapshot[counter] = openStore
+		counter++
+	}
+	p.lock.RUnlock()
+
+	for _, openStore := range openStoresSnapshot {
+		err := openStore.Close()
+		if err != nil {
+			return fmt.Errorf(`failed to close open store with name "%s": %w`, openStore.name, err)
+		}
+	}
+
+	return nil
+}
```

Comment:
```
func (p *Provider) Close() error {
	p.lock.Lock()
        deffer p.lock.Unlock()

	for _, openStore := range p.dbs {
		if err := openStore.Close(); err != nil {
			return fmt.Errorf(`failed to close open store with name "%s": %w`, openStore.name, err)
		}
	}

	return nil
}
```

Reply:
Done!

Reply:
Actually, I can't do this - the reason is that each store's close call calls into Provider.removeStore, so that results in a deadlock. The reason for the openStoresSnapshot thing is to work around this. (I do a similar thing in most other providers)